### PR TITLE
Qt controls gui combined

### PR DESCRIPTION
**Pull request by ayambabu23**
Combined controls and GUI in PyQt.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,6 @@
+*.kdev4
```
Comment (lutetjeff) [inline .gitignore]: move this line below the comment on line 2

File: gui.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
+# gui.py
+# Uses PyQt to collect camera footage and display information from Microcontroller
+# Uses PyQt to collect keyboard input too
+
+import random
+import sys
+
```
Comment (priime0) [inline gui.py]: Style: two blank lines between imports.

File: gui.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        self.thruster2speed: int = 0
+        self.thruster3speed: int = 0
+        self.thruster4speed: int = 0
+        self.servospeed: int = 0
+        
+        self.comms: Communications = comms
+        self.exit_program: ExitProgram = exit_program
+        
+        # (key, message sent to comms)
+        self.keys_pressed = [
```
Comment (lutetjeff) [inline gui.py]: probably better as a file-scope constant.

File: gui.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            ("D", "ld"),
+            ("d", "d"),
+            ("E", "le"),
+            ("e", "e"),
+            ("Q", "lq"),
+            ("q", "q"),
+            (" ", "spacebar")
+        ]
+        
+        self.keys_released = [
```
Comment (lutetjeff) [inline gui.py]: same here, probably better as a file-scope constant.

File: gui.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+sys.path.insert(0, './mcu-lib')
+app = QT.QApplication(sys.argv)
+
+
+class MainWindow(QT.QWidget):
+    def __init__(self, mcuobject: MCUInterface, comms: Communications, exit_program: ExitProgram):
+        super().__init__()
+        self.mcu: MCUInterface = mcuobject
+        self.thruster1speed: int = 0
```
Comment (lutetjeff) [inline gui.py]: maybe replace this with a `thruster_speed: list[int]`

File: gui.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        if not keyevent.isAutoRepeat():
+            for (key, trigger: str) in self.keys_released:
+                if keyevent.text() == key:
+                    self.on_trigger(trigger)
+                    
+    def closeEvent(self, QCloseEvent):
+        self.exit_program.Exit()
+
+   
+#feather = MCUInterface("/dev/ttyACM0")
```
Comment (priime0) [inline gui.py]: Remove unnecessary comments here.

File: serialtext.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import serial
```
Comment (priime0) [inline serialtext.py]: Is there a need to commit this file?

File: gui.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
+# gui.py
```
Comment (priime0) [inline gui.py]: Move this file into the `controls-pyqt` directory.

File: gui.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        self.z_accel.setText(str(self.mcu.latest_accel[2]))
+        self.temperature.setText(str(self.mcu.latest_temp))
+        self.thruster1.setText(str(self.thruster1speed))
+        self.thruster2.setText(str(self.thruster2speed))
+        self.thruster3.setText(str(self.thruster3speed))
+        self.thruster4.setText(str(self.thruster4speed))
+        self.servo.setText(str(self.servospeed))
+    
+    def setupui(self):
+        TITLE: str = 'MATE'
```
Comment (lutetjeff) [inline gui.py]: please, all constants belong on the file scope, not a class method!

File: gui.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        X_POSITION: int = 0
+        Y_POSITION: int = 0
+        LENGTH: int = 1600
+        WIDTH: int = 800
+        
+        self.setWindowTitle(TITLE)
```
Comment (lutetjeff) [inline gui.py]: modularize this chunk, between lines 81 and 147. use private methods (`__method`)

File: gui.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+            ("w", "sw"),
+            ("a", "sa"),
+            ("s", "ss"),
+            ("d", "sd"),
+            ("e", "se"),
+            ("q", "sq")
+        ]
+    
+    def updatetext(self):
+        self.voltage_info.setText(str(self.mcu.latest_voltage))
```
Comment (priime0) [inline gui.py]: Create variables for `self.mcu.latest_voltage`, `self.mcu.latest_gyro`, etc. and use those.

File: gui.py (hunk 10 above)
Comment (priime0) [inline gui.py]: For each block in this function, create a private function (prefix with `__`) whose name summarizes what the block of code does.

File: gui.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+                if keyevent.text() == key:
+                    self.on_trigger(trigger)
+                
+    def keyReleaseEvent(self, keyevent):
+        if not keyevent.isAutoRepeat():
+            for (key, trigger: str) in self.keys_released:
+                if keyevent.text() == key:
+                    self.on_trigger(trigger)
+                    
+    def closeEvent(self, QCloseEvent):
```
Comment (lutetjeff) [inline gui.py]: clean up; add `mcu.close_serial()` to avoid port blockage

File: gui.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+        self.keys_released = [
+            ("w", "sw"),
+            ("a", "sa"),
+            ("s", "ss"),
+            ("d", "sd"),
+            ("e", "se"),
+            ("q", "sq")
+        ]
+    
+    def updatetext(self):
```
Comment (priime0) [inline gui.py]: Change to `update_text`. If it is also a private function, prefix it with `__`.

File: serialtext.py (hunk 7 above)
Comment (lutetjeff) [inline serialtext.py]: What is this file for?

File: gui.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+        self.y_accel.setText(str(self.mcu.latest_accel[1]))
+        self.z_accel.setText(str(self.mcu.latest_accel[2]))
+        self.temperature.setText(str(self.mcu.latest_temp))
+        self.thruster1.setText(str(self.thruster1speed))
+        self.thruster2.setText(str(self.thruster2speed))
+        self.thruster3.setText(str(self.thruster3speed))
+        self.thruster4.setText(str(self.thruster4speed))
+        self.servo.setText(str(self.servospeed))
+    
+    def setupui(self):
```
Comment (priime0) [inline gui.py]: Change to `setup_ui`.

**Review by lutetjeff — CHANGES_REQUESTED** [review]
Mostly just formatting changes, but make sure you have docstrings on each method that should be used or seen, and private (`__`) all other methods.

File: gui.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+        TIMEOUT_INTERVAL = 100
+        
+        self.timer = QTimer()
+        self.timer.timeout.connect(self.updatetext)
+        self.timer.start(TIMEOUT_INTERVAL)
+    
+    def on_trigger(self, trigger: str):
+        self.comms.read_send(trigger)
+    
+    def keyPressEvent(self, keyevent):
```
Comment (priime0) [inline gui.py]: Use `snake_case` for `keyPressEvent`, `keyReleaseEvent`, and `closeEvent`.

**Review by priime0 — CHANGES_REQUESTED** [review]

File: controls-pyqt/exit_program.py [hunk 16]
```diff
@@ -0,0 +1,12 @@
+import sys
+
+
+class Exit_Program:
+    def __init__(self, comm: Communications, mcu: MCUInterface):
+        self.comms: Communications = comm
+        self.mcu: MCUInterface = mcu
+        
+    def Exit(self):
+        self.comms.kill_op()
```
Comment (priime0) [inline controls-pyqt/exit_program.py:9]: Would `comms.kill_op()` include `mcu.close_serial()`?

Comment (ayambabu23) [reply controls-pyqt/exit_program.py:9]: It can.

**Review by priime0 — APPROVED** [review]
lgtm 👍 . Wait for @lutet88 to review before the merge.

**Review by lutetjeff — APPROVED** [review]
Not quite ideal, but totally ready to merge.

**Pull request merged**